docs/src/design/stack/op_constraints.md

```diff
@@ -165,42 +167,42 @@ The degree of this flag is $3$, which is acceptable for a selector for degree $5
+* Constraints for the `U32SPLIT` operation have degree $3$. Thus, the degree of the op flag for this operation cannot exceed $6$.
+* Operations `U32ADD3` and `U32MADD` shift the stack to the left. Thus, having these two operations in this group and putting them under the common prefix `10011` allows us to create a common flag for these operations of degree $5$ (recall that the left-shift flag cannot exceed degree $5$).
 
 ### High-degree operations
-This group contains operations which require constraints with degree up to $3$. Similar to the previous group, the last op bit is not used in the computation of flag values for these operations.
+This group contains operations which require constraints with degree up to $3$. All $7$ operation bits are used for these flags. The extra $e_0$ column is used for degree reduction of the three high-degree bits.
 
 | Operation    | Opcode value | Binary encoding | Operation group                        | Flag degree |
 | ------------ | :----------: | :-------------: | :-------------------------------------:| :---------: |
-| `HPERM`      | $80$         | `101_0000`      | [Crypto ops](./crypto_ops.md)          | $6$         |
-| `MPVERIFY`   | $82$         | `101_0010`      | [Crypto ops](./crypto_ops.md)          | $6$         |
-| `PIPE`       | $84$         | `101_0100`      | [I/O ops](./io_ops.md)                 | $6$         |
-| `MSTREAM`    | $86$         | `101_0110`      | [I/O ops](./io_ops.md)                 | $6$         |
-| `SPAN`       | $88$         | `101_1000`      | [Flow control ops](../decoder/main.md) | $6$         |
-| `JOIN`       | $90$         | `101_1010`      | [Flow control ops](../decoder/main.md) | $6$         |
-| `SPLIT`      | $92$         | `101_1100`      | [Flow control ops](../decoder/main.md) | $6$         |
-| `LOOP`       | $94$         | `101_1110`      | [Flow control ops](../decoder/main.md) | $6$         |
+| `HPERM`      | $80$         | `101_0000`      | [Crypto ops](./crypto_ops.md)          | $5$         |
+| `MPVERIFY`   | $81$         | `101_0001`      | [Crypto ops](./crypto_ops.md)          | $5$         |
+| `PIPE`       | $82$         | `101_0010`      | [I/O ops](./io_ops.md)                 | $5$         |
+| `MSTREAM`    | $83$         | `101_0011`      | [I/O ops](./io_ops.md)                 | $5$         |
+| `SPLIT`      | $84$         | `101_0100`      | [Flow control ops](../decoder/main.md) | $5$         |
+| `LOOP`       | $85$         | `101_0101`      | [Flow control ops](../decoder/main.md) | $5$         |
+| `SPAN`       | $86$         | `101_0110`      | [Flow control ops](../decoder/main.md) | $5$         |
+| `JOIN`       | $87$         | `101_0111`      | [Flow control ops](../decoder/main.md) | $5$         |
```

Comment:
Should we add `unused` slots here (similar to how we have it on line 101 above)?